keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,364 @@
+The [Go implementation of OpenTelemetry](https://github.com/open-telemetry/opentelemetry-go) will be used.
+An [OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/otlptracegrpc/exporter.go) and
+an [OTLP trace provider](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/provider.go) along with a [context propagator](https://opentelemetry.io/docs/go/instrumentation/#propagators-and-context) will be configured.
+
+Also, designated http servers and clients will be wrapped with [otelhttp](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/v0.21.0/instrumentation/net/http)
+to generate spans for sampled incoming requests and propagate context with client requests.
+
+OpenTelemetry-Go provides the [propagation package](https://github.com/open-telemetry/opentelemetry-go/blob/main/propagation/propagation.go) with which you can add custom key-value pairs known as [baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md). Baggage data will be propagated across services within contexts.
+
+### Running the OpenTelemetry Collector
+
+The [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector) can be run as a sidecar, a daemonset, a deployment , or a combination in which the daemonset buffers telemetry and forwards to the deployment for aggregation (e.g. tail-base sampling) and routing to a telemetry backend.  To support these various setups, the kubelet should be able to send traffic either to a local (on the control plane network) collector, or to a cluster service (on the cluster network).
+
+### Kubelet Configuration
+
+Refer to [OpenTelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/ccbfdd9020ba0d581aa5d880235302f8b0eb8669/specification/resource/semantic_conventions/README.md#service)
+
+```golang
+// KubeletConfiguration contains the configuration for the Kubelet
+type KubeletConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
+    ------------
+
+    // +optional
+    // TracingConfiguration
+    TracingConfiguration TracingConfiguration
+}
+
+// TracingConfiguration specifies configuration for opentelemetry tracing
+type TracingConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // +optional
+    // URL of the collector that's running on the control-plane node.
+    // Defaults to 0.0.0.0:4317
+    URL  *string `json:"url,omitempty" protobuf:"bytes,1,opt,name=url"`
+
+    // +optional
+    // SamplingRatePerMillion is the number of samples to collect per million spans.
+    // Defaults to 0.
+    SamplingRatePerMillion *int32 `json:"samplingRatePerMillion,omitempty" protobuf:"varint,2,opt,name=samplingRatePerMillion"`
+}
+
+```
+If --opentelemetry-config-file is not specified, the kubelet will not send any spans, even if incoming requests ask for sampling.
```

Comment:
If this is part of the KubeletConfiguration, do we still need a flag?  If you specify the flag, does it take the TracingConfiguration, or KubeletConfiguration as input?

Would TracingConfiguration be part of ComponentBase as we've discussed, or specific to KubeletConfiguration?

Reply:
I was thinking this would be like the APIServer, where if the feature gate is enabled and if there is a tracing-config-file then spans will be generated. The tracing-config-file will contain the TracingConfiguration same as APIServer. 

Now I am remembering you suggested the ComponentBase, though.   I see the LoggingConfiguration and https://github.com/kubernetes/component-base/blob/master/logs/options.go. This would be similar, right? A TracingConfiguration & tracing options would be added to component-base. Then kubelet, scheduler, other core components would use this. And the APIServer tracing feature would be updated to use the TracingConfiguration from component-base? 

Thanks for the suggestion, this seems a good way forward since the goal is to instrument everything - you are more familiar with this,  so I leave it up to you but it makes sense to me.

Reply:
(I went ahead and updated this proposal to include reference to a new type in component-base)

Reply:
I'm pretty sure component-base is what we want to use, but we need to double-check that it will work here.  I'm not familiar with the interaction between KubeletConfiguration and component-base config (is it embedded in KubeletConfiguration, or is it a separate configuration?  Does the kubelet even use component-base at all?).  Otherwise, I think having it in ComponentBase will make future integrations much easier.

Reply:
And yes, I'd like the apiserver config to be transitioned to component-base if we go that route.

Reply:
Yup, KubeletConfiguration includes the componentbaseconfig.LoggingConfiguration